enhancements/substitutes-for.md

```diff
@@ -0,0 +1,430 @@
+
+Because the set of substitutions may grow over time, and are not guaranteed to be added in any particular 
+order, the substitutions must be applied in an order independent way.  This means each time a new 
+substitution rule is added, additional substitution rules must be generated to account for transitive
+substitutions.
+
+For example, suppose we have the following set of relationships we need to add to the catalog
+(In this example, A has a CVE, but B has already shipped as a replacement for A.  We want to patch
+A as it is still the head of some channel.  We produce A' to patch A, and then find another issue in
+A' and produce A'' to patch A'):
+
+B replaces A (original CSV relationship)
+A' substitutes for A
+A'' substitutes for A'
+
+We would first add the rule:
+A' -> A
+
+Since this is the first entry, there are no additional rules to generate.
+
+Next we add:
+A'' -> A'
+
+Which generates the following transitive substitution(A''->A'->A):
+A'' -> A
+
+So the total set of substitution rules is:
+A' -> A
+A'' -> A'
+A'' -> A
+
+Which means that ultimately B will be able to replace A, A', or A'', as desired.
+
+
+
+If we work the example in a different order we would first add the rule:
+A'' -> A'
+
+Since this is the first entry, there are no additional rules to generate.
+
+Next we add:
+A' -> A
+
+Which generates the following transitive substitution (A''->A'->A):
+A'' -> A
+
+Giving us the same total set of rules:
+A' -> A
+A'' -> A'
+A'' -> A
+
+Note that each time a new rule is added, the substitutions must be re-processed for additional
+new rules, until no new rules are generated.
+
+Given the set of substitution relationships:
+A' -> A
+A'' -> A'
+A'' -> A
+
+As well as the replaces definition in the CSV (B replaces A), it remains to
+to produce a catalog DB that will allow OLM to upgrade versions A'', A', and A' of the operator, 
+to version B.
+
+To achieve this, the existing bundle metadata in the catalog must be mutated as part of the opm 
+add processing.  Since "replaces" can only have a single entry, we will also need to add additional 
+"skips" entries to the bundle to account for the other valid upgrade paths.  Specifically, the mutated
+bundle for "B" will indicate that it replaces A''(the final substitution for A), and
+skips A and A' (the other values in the substitution relationship)
+
+In addition, the mutated version of A'' will add a skip for A.  (A'' will already skip A' as 
+provided by the CVE rebuilder).  Finally, A' will also skip A, as defined by the CVE rebuilder.
+
+This means the bundle metadata in the catalog DB will reflect that:
+
+B: replaces A''(originally was A), skips A, A'(result of substitution of A)
+A: whatever the original CSV listed
+A': skips A(added by the CVE rebuilder), plus original CSV
+A'': skips A'(added by the CVE rebuilder)+A(carried from CSV for A' when A' was rebuilt)+original CSV
+
+This in turn will be reported via ListBundles, ensuring that OLM can resolve the upgrade
```

Comment:
Quick nit: Depending on how far back this is backported, we may need to mutate other api calls instead of just ListBundles (otherwise, this change isn't actually backwards compatible on older clusters)

Reply:
intent here is that we don't mutate any calls, we mutate the data in the DB, so the api serving should just work.

This sentence was just intended to reflect that, once the DB content is mutated(on bundle add), the api results will reflect those mutations (as they would any DB changes).